### ISSUE TITLE
How to request an authentication token discussion (grant_types)

### ISSUE THREAD

**Issue by mcginkel** [stichtingsem/ecosystem#43]
in https://github.com/stichtingsem/ecosystem/pull/20 was a discussion started about how we will request an authentication token within the context of a certain school (based on digideliveryId)
This discussion was not finished, but PR was closed.

The normal way of using oauth2 with client_credentials is supported bij standard libraries, but does not offer a way to add a digideliveryId to the request. The authentication server needs this to see if there is consent for that client to process data on behalf of the school. If consent was given bij the school administrator the authentication token is given and used in allsubsequent requests for this schools data. The digideliveryId is somehow added as claims in the token. How is not important.

There are different ways to support this:

**Comment by mcginkel** [comment]
@Jorimvdw 
I agree with Henk on this one. If it limits the libraries which we can use, we shouldnt make it custom

**Comment by mcginkel** [comment]
@Arjen Post: (iddink)
The suggested client credentials flow is [defined in the OAuth 2.0 RFC](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.4). The parameter acr_values is [defined in the OpenID Connect Core specification](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) as a parameter for an authorize request in the authorization code flow (section 3.1.2.1). The parameter acr_values could be used in the client credentials flow, but is far from standard.

Besides that, I can imagine that adding acr_values to the client credentials flow could lead to comparable limitations in OAuth 2.0 libraries.

And the client credentials flow is so simple you could wonder if you really need an OAuth library for that. Auth0 has some sample implementations of a client credentials token request in multiple languages: https://auth0.com/docs/get-started/authentication-and-authorization-flow/call-your-api-using-the-client-credentials-flow.

**Comment by hhaarsma** [comment]
I brought this discussion under attention by one of the engineers at our place who is most experienced with authentication mechanisms, he sees 3 options for putting the school information in the JWT token:
- using a custom grant - this would not be his prevered solution in this scenario
- using Client/Secret with a custom parameter
- using a parameterized scope, where first part of the requested scope is fixed and the second part is dynamic (so in our case the school reference) - I dont know to what extend this is a standard, and how much effort this takes in other products (here is a description of the approach that could be used: https://docs.identityserver.io/en/latest/topics/resources.html#parameterized-scopes) 
He indicated that he would prever the second or the last option (or to pass it as a parameter in each API call, but that option has already been discussed, and is not considered in this discussion).


During my duscussion with him, I started wondering whether the JWT token can be cached if consent validation is done before handing over the JWT token (so that the API doesnt have to do it) - this might give strange effects in the proces where a consent is communicated between 2 partners, and then the initialization flow is started - then it might happen that during the initialization flow a cached token is used in which the consent has not been taken into account.

**Comment by mcginkel** [comment]
some hints on adding parameters or changing grant types if you use spring boot java: https://www.baeldung.com/spring-security-custom-oauth-requests

**Comment by mcginkel** [comment]
@hhaarsma About caching.
I expect you can use caching of tokens, provided that you request data for the same scopes and school.

**Comment by mcginkel** [comment]
For c# : https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.openidconnect.openidconnectevents.onredirecttoidentityprovider?view=aspnetcore-6.0
this is the point to modify a normal request and add the schoolid and/or change grand type using normal libraries

**Comment by Jorimvdw** [comment]
I was just thinking; shouldnt we also add/write down the scope(s) in the token?

**Comment by hhaarsma** [comment]
I think we will implement the Consent check on the service call it self (to prevent calls being done with JWT tokens which contain old consent information (either consent has been given in the mean time, or consent has been redrawn)). 
In the normal flows the requested scopes for client/secret are to request acces for certain end-points/functionalities - to me it feels that in our situation we replace that mechanism with our custom consents.

For example say party X requests a token with the scope to post SIS data, then on the client of party X that scope should be configured as being allowed; but then still the consent for the school in that request has to be checked. Since we want to create an open eco-system I dont expect that we will determine that parties X and Y are allowed to do something with SIS data (if the school has given consent), but party Z is not allowed to do so (even with consent of the school).

**Comment by Jorimvdw** [comment]
Few remarks:
- Using Client/Secret with a custom parameter is our preferred solution
- The parameter should be optional (for certain calls it's not necessary, for example catalogue)
- I agree with Henk in concern to when to check for consent. Don't check it when a token is requested, check it when the api is called
- Question: scopes are out of the token and are registered with a client. Do they play any role when requesting a token?

**Comment by hhaarsma** [comment]
I dont think the scopes are needed in the sence of indicating what the intent of the token is and whether the client can be used for that (this is basically the authorisation part which will be done with the consents which are given). 
I can imagine that we use a scope to make clear that this is used in the SEM context and that the additional parameter check should be done.

**Comment by kvoer** [comment]
Discussion is addressed during the Technical Track meeting of 4th of March. @Jorimvdw adds a Pull Request based on the solution @hhaarsma suggested.
